# TP4. Regresssion

In [1]:
from __future__ import absolute_import, division, print_function
%matplotlib inline

import numpy as np
from numpy.linalg import inv
import matplotlib.pyplot as plt

We first generate data with 20 features; among them, 10 features are uncorrelated to y

In [2]:
def g(x): # some favorite function
    s = 0
    for i in range(len(x)):
        s += (i%2) * x[i] * np.tan(x[i]) * np.sqrt(i+1)
    return 10*s

np.random.seed(999)
a = np.arange(0,1,0.01)
X = np.random.choice(a, size= 2000, replace=True).reshape((100,20))

y = np.empty(len(X))
for i in range(len(X)):
    y[i] = g(X[i])

X_train = X[:50]
y_train = y[:50]
X_test = X[50:]
y_test = y[50:]

In [4]:
X_train

array([[0.64, 0.92, 0.97, 0.72, 0.97, 0.91, 0.31, 0.89, 0.51, 0.16, 0.08,
        0.08, 0.48, 0.69, 0.66, 0.11, 0.11, 0.21, 0.87, 0.5 ],
       [0.62, 0.71, 0.17, 0.62, 0.43, 0.32, 0.04, 0.13, 0.43, 0.84, 0.72,
        0.27, 0.5 , 0.23, 0.52, 0.84, 0.05, 0.43, 0.81, 0.95],
       [0.54, 0.42, 0.88, 0.57, 0.49, 0.43, 0.01, 0.78, 0.7 , 0.03, 0.72,
        0.09, 0.78, 0.16, 0.63, 0.49, 0.58, 0.31, 0.46, 0.52],
       [0.76, 0.14, 0.48, 0.06, 0.04, 0.87, 0.67, 0.52, 0.63, 0.25, 0.43,
        0.94, 0.27, 0.22, 0.78, 0.39, 0.3 , 0.51, 0.63, 0.48],
       [0.75, 0.52, 0.04, 0.54, 0.56, 0.38, 0.78, 0.76, 0.59, 0.48, 0.15,
        0.22, 0.71, 0.15, 0.49, 0.02, 0.21, 0.46, 0.11, 0.31],
       [0.6 , 0.58, 0.98, 0.33, 0.99, 0.35, 0.66, 0.74, 0.19, 0.35, 0.12,
        0.04, 0.01, 0.06, 0.37, 0.89, 0.11, 0.79, 0.02, 0.66],
       [0.83, 0.88, 0.8 , 0.48, 0.89, 0.16, 0.23, 0.16, 0.69, 0.77, 0.61,
        0.98, 0.94, 0.06, 0.08, 0.89, 0.84, 0.25, 0.5 , 0.05],
       [0.72, 0.14, 0.19, 0.67, 0.25, 0.8

### Linear Regression
Read the following documentation and understand the code

http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html

In [31]:
from sklearn.linear_model import LinearRegression

lin_reg = LinearRegression()
lin_reg.fit(X_train, y_train)

y_train_pred = lin_reg.predict(X_train)
mse_train = np.mean((y_train-y_train_pred)**2 )
print('MSE of train dataset:', mse_train)

y_pred = lin_reg.predict(X_test)
mse_test = np.mean((y_test-y_pred)**2 )
print('MSE of test dataset:', mse_test)

coefs = lin_reg.coef_
print('Coefficients:\n', coefs) # contain 20 coefficients, not the independent coefficient

MSE of train dataset: 120.19826684914514
MSE of test dataset: 356.4868075878598
Coefficients:
 [ 15.26455355  40.46800213 -10.21335792  27.4920002   -4.63610068
  40.66508376   3.47899568  32.47264792  -9.79806108  34.04082205
  -4.44214276  45.93739173  11.19814633  61.90038321  -1.29734991
  65.33791401  14.1195483   72.16802276  -2.85811332  57.54438325]


In [32]:
lin_reg.score(X_train, y_train)

0.9516418022075496

### Exercise. Linear regression with ridge regulation
Read the documentation and edit the code above to get Linear regression with Ridge regulation. Try different alpha to see how it affects the model. Which alpha should we choose?

http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.Ridge.html
    

In [22]:
from sklearn.linear_model import Ridge

# Your code here
rid_reg = Ridge(alpha=0.7)
rid_reg.fit(X_train, y_train)

y_train_pred = rid_reg.predict(X_train)
mse_train = np.mean((y_train-y_train_pred)**2 )
print('MSE of train dataset:', mse_train)

y_pred = rid_reg.predict(X_test)
mse_test = np.mean((y_test-y_pred)**2 )
print('MSE of test dataset:', mse_test)

coefs = rid_reg.coef_
print('Coefficients:\n', coefs) # contain 20 coefficients, not the independent coefficient

score_train = rid_reg.score(X_train, y_train)
score_test = rid_reg.score(X_test, y_test)

print(f'Train score: {score_train}\n', f'Test score: {score_test}')

MSE of train dataset: 169.7320573163789
MSE of test dataset: 300.460330778744
Coefficients:
 [  5.37054384  37.84749771  -8.99011113  22.27589868  -3.58637528
  26.6493153   -0.34966779  31.28105216 -11.73115988  26.40883294
   0.4454349   39.88660581  13.56874581  49.87785979   3.9557953
  57.06899624   8.49652793  58.47344108   0.24539891  45.10819341]
Train score: 0.9317133548212774
 Test score: 0.8384773328175237


With alpha 1: 
Train score: 0.9175810801161374 
Test score: 0.8319105110902633

alpha 0.5:
Train score: 0.9400103991439156
 Test score: 0.8389792067849651
 
alpha 0.2:
Train score: 0.9492855477526201
 Test score: 0.8288187223278736
 
alpha 0.1:
Train score: 0.9509966901433983
 Test score: 0.8205503942875743
 
 
alpha 0.5 is the best (for test)

### Exercise. Linear Regression with Lasso regulation
The same question as Exercise 3, with lasso.

http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.Lasso.html

In [30]:
from sklearn.linear_model import Lasso

# Your code here
reg = Lasso(alpha=0.1)
reg.fit(X_train, y_train)

y_train_pred = reg.predict(X_train)
mse_train = np.mean((y_train-y_train_pred)**2 )
print('MSE of train dataset:', mse_train)

y_pred = reg.predict(X_test)
mse_test = np.mean((y_test-y_pred)**2 )
print('MSE of test dataset:', mse_test)

coefs = reg.coef_
print('Coefficients:\n', coefs) # contain 20 coefficients, not the independent coefficient

score_train = reg.score(X_train, y_train)
score_test = reg.score(X_test, y_test)

print(f'Train score: {score_train}\n', f'Test score: {score_test}')

MSE of train dataset: 123.94206644278037
MSE of test dataset: 326.69112385728096
Coefficients:
 [11.54630312 40.79030357 -8.58883024 25.50053725 -2.56809473 36.90869972
  1.84081467 31.3209008  -7.55390147 31.32409054 -1.30193042 44.61384862
 10.61735501 60.59198544  0.         64.32910847 11.33198413 69.43211171
 -1.19278812 55.20251859]
Train score: 0.9501355957871896
 Test score: 0.8243760780882365


With alpha 1: 
Train score: 0.8867234144723495
 Test score: 0.7669682024697592
 
alpha 0.5:
Train score: 0.9285132172062385
 Test score: 0.8316585840643945
 
alpha 0.2:
Train score: 0.9462787521650388
 Test score: 0.8342725340181228

alpha 0.1:
Train score: 0.9501355957871896
 Test score: 0.8243760780882365
 
alpha 0.2 is best choices

## Polynomial regression

http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.PolynomialFeatures.html#sklearn.preprocessing.PolynomialFeatures

## Exercise
Try to do Polynomial Regression with this data

In [44]:
from sklearn.preprocessing import PolynomialFeatures

# Your code here
pol_pre = PolynomialFeatures(degree=2)
X_train_poly = pol_pre.fit_transform(X_train)
X_test_poly = pol_pre.fit_transform(X_test)

lin_reg = LinearRegression()
lin_reg.fit(X_train_poly, y_train)

y_train_pred = lin_reg.predict(X_train_poly)
mse_train = np.mean((y_train-y_train_pred)**2 )
print('MSE of train dataset:', mse_train)

y_pred = lin_reg.predict(X_test_poly)
mse_test = np.mean((y_test-y_pred)**2 )
print('MSE of test dataset:', mse_test)

coefs = lin_reg.coef_
print('Coefficients:\n', coefs) 

MSE of train dataset: 1.1070810834999262e-26
MSE of test dataset: 335.4725862555032
Coefficients:
 [  0.23878316   0.08497305   4.50752602   0.97766033   5.5358902
  -2.54075115   4.94670491  -6.94789545  -1.03532451  -5.14679153
  -0.96442691  -3.52805289   3.50669789  -3.12226864   8.5037629
  -1.18954345  11.82104386   4.39927852   6.11946868   0.36077221
   8.61168513   7.31955863   2.63664178   5.09795969  -3.28395981
   3.08933381  -0.22270594  -3.3968908   -5.04588285  -1.856867
  12.81304423   2.88057135   7.82614821  -3.70908313  -3.45405888
  -1.97145273   1.4909435   -4.42013715   2.68105208  -2.80986297
   5.0726264    6.18758546   0.93407666  -3.32739339   7.95976819
   3.47295044   2.12715223   2.63725748   2.68736269  13.90284749
   3.91721584   0.80325992  -2.90887685  -1.82288956   8.75673251
   3.733771    -2.40481475   7.76580325   2.58309373   1.34473181
   1.68230074  -5.06410154  -2.5887426    1.47173176   0.25873745
   5.61101015  -1.78389676  -2.7685393   -4.778

In [48]:
score_train = lin_reg.score(X_train_poly, y_train)
score_test = lin_reg.score(X_test_poly, y_test)

print(f'Train score: {score_train}\n', f'Test score: {score_test}')

Train score: 1.0
 Test score: 0.8196553043852749


### Exercise:

Try the following models in sklearn:
- Support Vector Regressor
- Decision Tree Regressor
- Multi-layer Perceptron Regressor (Basic Neural Network)


In [ ]:
# Saving datas
temps = [X_train, X_test, y_train, y_test]

#### Support Vector

In [51]:
from sklearn.svm import SVR

# Instanciate regressor
clf = SVR(C=1.0, epsilon=0.2)

clf.fit(X_train, y_train)

y_train_pred = clf.predict(X_train)
mse_train = np.mean((y_train-y_train_pred)**2 )
print('MSE of train dataset:', mse_train)

y_test_pred = clf.predict(X_test)
mse_test = np.mean((y_test-y_pred)**2 )
print('MSE of test dataset:', mse_test)

# coefs = clf.coef_
# print('Coefficients:\n', coefs) 

score_train = clf.score(X_train, y_train)
score_test = clf.score(X_test, y_test)

print(f'Train score: {score_train}\n', f'Test score: {score_test}')

MSE of train dataset: 2373.8196544876273
MSE of test dataset: 1823.0074114975782
Train score: 0.04496426292576361
 Test score: 0.019980379322200315


#### Decision Tree

In [59]:
from sklearn.tree import DecisionTreeRegressor

# Instanciate regressor
reg = DecisionTreeRegressor(random_state=1)

reg.fit(X_train, y_train)

y_train_pred = reg.predict(X_train)
mse_train = np.mean((y_train-y_train_pred)**2 )
print('MSE of train dataset:', mse_train)

y_test_pred = reg.predict(X_test)
mse_test = np.mean((y_test-y_test_pred)**2 )
print('MSE of test dataset:', mse_test)

# coefs = clf.coef_
# print('Coefficients:\n', coefs) 

score_train = reg.score(X_train, y_train)
score_test = reg.score(X_test, y_test)

print(f'Train score: {score_train}\n', f'Test score: {score_test}')

MSE of train dataset: 0.0
MSE of test dataset: 2869.6717844779105
Train score: 1.0
 Test score: -0.542689643474092


#### Multi-layer Perceptron Regressor 

In [60]:
from sklearn.neural_network import MLPRegressor

# Instanciate regressor
reg = MLPRegressor(alpha=1)

reg.fit(X_train, y_train)

y_train_pred = reg.predict(X_train)
mse_train = np.mean((y_train-y_train_pred)**2 )
print('MSE of train dataset:', mse_train)

y_test_pred = reg.predict(X_test)
mse_test = np.mean((y_test-y_test_pred)**2 )
print('MSE of test dataset:', mse_test)

# coefs = clf.coef_
# print('Coefficients:\n', coefs) 

score_train = reg.score(X_train, y_train)
score_test = reg.score(X_test, y_test)

print(f'Train score: {score_train}\n', f'Test score: {score_test}')

MSE of train dataset: 8920.657780278954
MSE of test dataset: 6887.331492721275
Train score: -2.588961344333725
 Test score: -2.702519229712926


/home/matt/anaconda3/lib/python3.7/site-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)
